pkg/monitor/cluster/healthz.go

```diff
@@ -5,17 +5,53 @@ package cluster
 
 import (
 	"strconv"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/errors"
 )
 
 func (mon *Monitor) emitAPIServerHealthzCode() (int, error) {
 	var statusCode int
-	err := mon.cli.Discovery().RESTClient().
+	raw, err := mon.cli.Discovery().RESTClient().
 		Get().
-		AbsPath("/healthz").
+		AbsPath("/readyz").
 		Do().
 		StatusCode(&statusCode).
-		Error()
+		Raw()
+
+	// HTTP 500s are expected behaviour (e.g. on startup/shutdown of the API server) and should not cause this function to return an error. Therefore, silence it.
+	if err != nil && errors.IsUnexpectedServerError(err) {
+		err = nil
+	}
```

Comment:
I don't like the idea of silencing 500 alerts -- while they are expected sometimes, is it guaranteed that they always are?

I've been looking for a list of all possible conditions the `/healthz` and `/readyz` endpoints can emit from the KAS, without much success. Do you know if this list exists somewhere? Wondering where you got this info. :)

Reply:
I looked at the code! :) https://github.com/kubernetes/kubernetes/blob/release-1.16/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go#L162-L206

There's three exits -- one marking the request as an error and returning the verbose results, one returning normally, and one returning normally but giving the verbose results if it was requested.

Reply:
Yeah that's what I was afraid of -- this is liable to change between versions of k8s, and also, I think it's still possible for the underlying webserver to return a different response code if it never reaches this code due to an earlier panic, right?

I only want to rely on such behaviour if it's a documented contract. In my experience (*puts on SIG Chair :womans_hat:*) Kubernetes does not provide any compatibility guarantees on response codes or metrics, with very limited exceptions (e.g. stable metrics, which still don't actually exist).